FFmpegInterop/FFmpegInteropBuffer.cpp

```diff
@@ -0,0 +1,184 @@
+		m_length(static_cast<uint32_t>(buf.size()))
+	{
+		// We need to move capture buf in the deleter lambda before setting m_buf, 
+		// since moving buf invalidates the pointer returned by buf.data().
+		uint8_t* p{ nullptr };
+		function<void(uint8_t*)> deleter = [pp = &p, buf = move(buf)](uint8_t*) mutable noexcept
+		{
+			if (pp != nullptr)
+			{
+				*pp = const_cast<uint8_t*>(buf.data());
+				pp = nullptr;
+			}
+		};
+
+		// Call lambda to set p
+		deleter(nullptr);
+
+		m_buf.reset(p);
+		m_buf.get_deleter() = move(deleter);
+	}
+
+	uint32_t FFmpegInteropBuffer::Capacity() const noexcept
+	{
+		return m_length;
+	}
+
+	uint32_t FFmpegInteropBuffer::Length() const noexcept
+	{
+		return m_length;
+	}
+
+	void FFmpegInteropBuffer::Length(_In_ uint32_t length)
+	{
+		// We have a fixed length
+		THROW_HR(CO_E_NOT_SUPPORTED);
+	}
+
+	STDMETHODIMP FFmpegInteropBuffer::Buffer(_Outptr_ uint8_t** buf) noexcept
+	{
+		RETURN_HR_IF_NULL(E_POINTER, buf);
+		*buf = m_buf.get();
+		return S_OK;
+	}
+
+	STDMETHODIMP FFmpegInteropBuffer::GetUnmarshalClass(
+		_In_ REFIID riid,
+		_In_opt_ void* pv,
+		_In_ DWORD dwDestContext,
+		_Reserved_ void* pvDestContext,
+		_In_ DWORD mshlflags,
+		_Out_ CLSID* pclsid) noexcept
+	{
+		RETURN_IF_FAILED(InitMarshaler());
+		RETURN_IF_FAILED(m_marshaler->GetUnmarshalClass(riid, pv, dwDestContext, pvDestContext, mshlflags, pclsid));
+		return S_OK;
+	}
+
+	STDMETHODIMP FFmpegInteropBuffer::GetMarshalSizeMax(
+		_In_ REFIID riid,
+		_In_opt_ void* pv,
+		_In_ DWORD dwDestContext,
+		_Reserved_ void* pvDestContext,
+		_In_ DWORD mshlflags,
+		_Out_ DWORD* pcbSize) noexcept
+	{
+		RETURN_IF_FAILED(InitMarshaler());
+		RETURN_IF_FAILED(m_marshaler->GetMarshalSizeMax(riid, pv, dwDestContext, pvDestContext, mshlflags, pcbSize));
+		return S_OK;
+	}
+
+	STDMETHODIMP FFmpegInteropBuffer::MarshalInterface(
+		_In_ IStream* pStm,
+		_In_ REFIID riid,
+		_In_opt_ void* pv,
+		_In_ DWORD dwDestContext,
+		_Reserved_ void* pvDestContext,
+		_In_ DWORD mshlflags) noexcept
+	{
+		RETURN_IF_FAILED(InitMarshaler());
+		RETURN_IF_FAILED(m_marshaler->MarshalInterface(pStm, riid, pv, dwDestContext, pvDestContext, mshlflags));
+		return S_OK;
+	}
+
+	STDMETHODIMP FFmpegInteropBuffer::UnmarshalInterface(_In_ IStream* pStm, _In_ REFIID riid, _Outptr_ void** ppv) noexcept
+	{
+		RETURN_IF_FAILED(InitMarshaler());
+		RETURN_IF_FAILED(m_marshaler->UnmarshalInterface(pStm, riid, ppv));
+		return S_OK;
+	}
+
+	STDMETHODIMP FFmpegInteropBuffer::ReleaseMarshalData(_In_ IStream* pStm) noexcept
+	{
+		RETURN_IF_FAILED(InitMarshaler());
+		RETURN_IF_FAILED(m_marshaler->ReleaseMarshalData(pStm));
+		return S_OK;
+	}
+
+	STDMETHODIMP FFmpegInteropBuffer::DisconnectObject(_Reserved_ DWORD dwReserved) noexcept
+	{
+		RETURN_IF_FAILED(InitMarshaler());
+		RETURN_IF_FAILED(m_marshaler->DisconnectObject(dwReserved));
+		return S_OK;
+	}
+
+	HRESULT FFmpegInteropBuffer::InitMarshaler() noexcept
+	{
+		if (m_marshaler == nullptr)

```

Comment:
>m_marshaler == nullptr [](start = 6, length = 22)

You will need to use InterlockedCompareExchangePointer here to ensure that this read is synchronized with the write within the if statement (i.e InterlockedCompareExchangePointer(m_marshaler.addressof(), nullptr, nullptr) == nullptr). Otherwise, on some platforms, this check could evaluate to false while another thread is still writing to m_marshaler, causing a subsequent method call on an invalid m_marshaler pointer.

Reply:
Nice catch! I modeled this code after other [Microsoft sample projects](https://github.com/search?q=org%3Amicrosoft+RoGetBufferMarshaler&type=Code) so we should fix this issue in those projects too.

I'm actually going to switch to using std::call_once() for lazy initialization here. Calling com_ptr::put_void() is technically valid here because we'll only update m_marshaler if it's null, however, this usage still triggers the assert in com_ptr::put() which is problematic on debug builds.